merkledag.go

```diff
@@ -281,30 +282,94 @@ func GetLinksWithDAG(ng ipld.NodeGetter) GetLinks {
+type WalkOptions struct {
+	WithRoot     bool
+	IgnoreErrors bool
+	Concurrency  int
+}
+
+// WalkOption is a setter for WalkOptions
+type WalkOption func(*WalkOptions)
+
+// WithRoot is a WalkOption indicating that the root node should be visited
+func WithRoot() WalkOption {
+	return func(walkOptions *WalkOptions) {
+		walkOptions.WithRoot = true
+	}
+}
+
+// Concurrent is a WalkOption indicating that node fetching should be done in
+// parallel, with the default concurrency factor.
+// NOTE: When using that option, the walk order is *not* guarantee.
+// NOTE: It *does not* make multiple concurrent calls to the passed `visit` function.
+func Concurrent() WalkOption {
+	return func(walkOptions *WalkOptions) {
+		walkOptions.Concurrency = defaultConcurrentFetch
+	}
+}
+
+// Concurrency is a WalkOption indicating that node fetching should be done in
+// parallel, with a specific concurrency factor.
+// NOTE: When using that option, the walk order is *not* guarantee.
+// NOTE: It *does not* make multiple concurrent calls to the passed `visit` function.
+func Concurrency(worker int) WalkOption {
+	return func(walkOptions *WalkOptions) {
+		walkOptions.Concurrency = worker
+	}
+}
+
+// IgnoreErrors is a WalkOption indicating that the walk should attempt to
+// continue even when an error occur.
+func IgnoreErrors() WalkOption {
+	return func(walkOptions *WalkOptions) {
+		walkOptions.IgnoreErrors = true
+	}
+}
+
 // WalkGraph will walk the dag in order (depth first) starting at the given root.
-func Walk(ctx context.Context, getLinks GetLinks, root cid.Cid, visit func(cid.Cid) bool) error {
+func Walk(ctx context.Context, getLinks GetLinks, c cid.Cid, visit func(cid.Cid) bool, options ...WalkOption) error {
 	visitDepth := func(c cid.Cid, depth int) bool {
 		return visit(c)
 	}
 
-	return WalkDepth(ctx, getLinks, root, 0, visitDepth)
+	return WalkDepth(ctx, getLinks, c, visitDepth, options...)
 }
 
 // WalkDepth walks the dag starting at the given root and passes the current
 // depth to a given visit function. The visit function can be used to limit DAG
 // exploration.
-func WalkDepth(ctx context.Context, getLinks GetLinks, root cid.Cid, depth int, visit func(cid.Cid, int) bool) error {
-	if !visit(root, depth) {
-		return nil
+func WalkDepth(ctx context.Context, getLinks GetLinks, c cid.Cid, visit func(cid.Cid, int) bool, options ...WalkOption) error {
+	opts := &WalkOptions{}
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	if opts.Concurrency > 1 {
+		return parallelWalkDepth(ctx, getLinks, c, visit, opts)
+	} else {
+		return sequentialWalkDepth(ctx, getLinks, c, 0, visit, opts)
+	}
+}
+
+func sequentialWalkDepth(ctx context.Context, getLinks GetLinks, root cid.Cid, depth int, visit func(cid.Cid, int) bool, options *WalkOptions) error {
+	if depth != 0 || options.WithRoot {
+		if !visit(root, depth) {
+			return nil
+		}
 	}
 
 	links, err := getLinks(ctx, root)
-	if err != nil {
+	if err != nil && !options.IgnoreErrors {
```

Comment:
So, the _idea_ is that `getLinks` should this internally. But I agree this option is convenient.

What if we changed this to an error handler (`func(cid.Cid, error) error`)? That way, we can provide options like:

```
IgnoreErrors()
IgnoreMissing() // only ignores ipld.ErrorNotFound.
OnMissing(func(c cid.Cid))
```

Reply:
Good idea, I'll add a generic `OnError(err error, c cid.Cid)` as well.